packages/client-benchmark/src/generateClient.ts

```diff
@@ -0,0 +1,110 @@
+import fs from 'fs/promises'
+import path from 'path'
+
+import { clearDir } from './fs'
+import { pnpmPrismaDbPush } from './pnpm'
+
+function getHeader(features: string[]) {
+  const featuresStr = JSON.stringify(['tracing', ...features])
+  return /* Prisma */ `
+generator client {
+  provider = "prisma-client-js"
+  output = "./client"
+  previewFeatures = ${featuresStr}
+}
+
+datasource db {
+  provider = "postgresql"
+  url = env("TEST_POSTGRES_URI")
+}`
+}
+
+class Model {
+  private fields: string[]
+  constructor(readonly name: string) {
+    this.fields = []
+  }
+
+  addField(field: string) {
+    this.fields.push(field)
+    return this
+  }
+
+  stringify() {
+    return `
+model ${this.name} {
+${this.fields.map((f) => `  ${f}`).join('\n')}
+}`
+  }
+}
+
+function addRelation(from: Model, to: Model, baseName: string) {
+  const kind = randomElement(['1:1', '1:m', 'm:n'])
+
+  if (kind === '1:1') {
+    from
+      .addField(`${baseName}Id String @unique`)
+      .addField(`${baseName}Forward ${to.name}? @relation("${baseName}", fields: [${baseName}Id], references: [id])`)
+
+    to.addField(`${baseName}Back ${from.name}? @relation("${baseName}")`)
+  } else if (kind === '1:m') {
+    from
+      .addField(`${baseName}Id String`)
+      .addField(`${baseName}Forward ${to.name} @relation("${baseName}", fields: [${baseName}Id], references: [id])`)
+
+    to.addField(`${baseName}Back ${from.name}[] @relation("${baseName}")`)
+  } else {
+    from.addField(`${baseName}m ${to.name}[] @relation("${baseName}")`)
+    to.addField(`${baseName}n ${from.name}[] @relation("${baseName}")`)
+  }
+}
+
+function genSchemaText(numModels: number, numRelations: number, features: string[]) {
+  const models = [
+    new Model('User')
+      .addField('id String @id @default(uuid())')
+      .addField('email String')
+      .addField('fieldA String')
+      .addField('fieldB Int')
+      .addField('fieldC DateTime'),
+  ]
+
+  for (let i = 0; i < numModels - 1; i++) {
+    models.push(
+      new Model(`Model${i}`)
+        .addField('id String @id @default(uuid())')
+        .addField('fieldA String')
+        .addField('fieldB Int')
+        .addField('fieldC DateTime'),
+    )
+  }
+
+  for (let i = 0; i < numRelations; i++) {
+    addRelation(randomElement(models), randomElement(models), `rel${i}`)
+  }
+
+  return [getHeader(features)].concat(models.map((m) => m.stringify())).join('\n')
+}
+
+function randomElement(array) {
+  return array[Math.floor(Math.random() * array.length)]
```

Comment:
I think the last element will never be chosen using this formula

Reply:
No, i think this is correct:

1. `Math.random()` is always in [0;1) range.
2. `Math.random() * array.length` therefore is in in [0;array.length) range
3. `Math.floor(Math.random() * array.length)` is then always in [0; array.length - 1] (inclusive) range.

Reply:
ah yes, you are right, sorry